examples/drawing.rs

```diff
@@ -0,0 +1,200 @@
+            set_vexpand: true,
+            set_hexpand: true,
+
+            add_controller = &gtk::GestureClick::new() {
+              set_button: 0,
+              connect_pressed[sender] => move |controller, _, x, y| {
+                if controller.current_button() == gtk::gdk::BUTTON_SECONDARY {
+                    sender.input(AppMsg::Reset);
+                } else {
+                    sender.input(AppMsg::AddPoint((x, y)));
+                }
+              }
+            },
+            connect_resize[sender] => move |_, x, y| {
+                sender.input(AppMsg::Resize((x, y)));
+            }
+          },
+        }
+      }
+    }
+
+    fn update(&mut self, msg: AppMsg, _sender: ComponentSender<Self>) {
+        self.reset = false;
+        match msg {
+            AppMsg::AddPoint((x, y)) => {
+                self.points.push(Point::new(x, y));
+            }
+            AppMsg::UpdatePoints => {
+                for point in &mut self.points {
+                    let Point { x, y, .. } = point;
+                    if *x < 0.0 {
+                        point.xs = point.xs.abs();
+                    } else if *x > self.width {
+                        point.xs = -point.xs.abs();
+                    }
+                    *x = x.clamp(0.0, self.width);
+                    *x += point.xs;
+
+                    if *y < 0.0 {
+                        point.ys = point.ys.abs();
+                    } else if *y > self.height {
+                        point.ys = -point.ys.abs();
+                    }
+                    *y = y.clamp(0.0, self.height);
+                    *y += point.ys;
+                }
+            }
+            AppMsg::Resize((x, y)) => {
+                self.width = x as f64;
+                self.height = y as f64;
+            }
+            AppMsg::Reset => {
+                self.reset = true;
+            }
+        }
+
+        let cx = self.handler.get_context().unwrap();
+        if self.reset {
+            cx.set_operator(Operator::Clear);
+            cx.set_source_rgba(0.0, 0.0, 0.0, 0.0);
+            cx.paint().expect("Couldn't fill context");
+        }
+        draw(&cx, &self.points);
+    }
+
+    fn update_cmd(&mut self, msg: Self::CommandOutput, sender: ComponentSender<Self>) {
+        sender.input(msg);
+    }
+
+    fn init(_params: (), root: &Self::Root, sender: ComponentSender<Self>) -> ComponentParts<Self> {
+        let mut model = AppModel {
+            width: 100.0,
+            height: 100.0,
+            points: Vec::new(),
+            handler: DrawHandler::new().unwrap(),
+            reset: false,
+        };
+
+        let widgets = view_output!();
+
+        model.handler.init(&widgets.area);
+
+        sender.command(|out, shutdown| {
+            shutdown
+                .register(async move {
+                    loop {
+                        tokio::time::sleep(Duration::from_millis(20)).await;
+                        out.send(AppMsg::UpdatePoints);
```

Comment:
You're normally not supposed to do so, but you could move a clone of the sender into the closure to send messages directly to input. Or even better, remove the AppMsg::UpdatePoints variant and update the points only from `update_cmd`.

Reply:
> You're normally not supposed to do so, but you could move a clone of the sender into the closure to send messages directly to input.

I'm a bit wary of doing something you're "not supposed to do" in exemplary code. Maybe it's fine with a comment?

> remove the AppMsg::UpdatePoints variant and update the points only from `update_cmd`.

I tried this, but then a view update isn't triggered.

Reply:
> I tried this, but then a view update isn't triggered.

Actually, `update_cmd` should trigger a view update.

Reply:
You're right, I was missing a call to `draw`. Fixed, and simplified the reset logic a bit, too.